contracts/mixins/ProtocolTokenHandler.sol

```diff
@@ -0,0 +1,138 @@
+    }
+
+    /**
+     * @param requiredCount_ required amount of signs provided by users
+     */
+    function setRequiredCount(uint256 requiredCount_) public onlyOwner {
+        require(requiredCount > 0, "required amount of signs should larger than zero");
+
+        requiredCount = requiredCount_;
+
+        emit SetRequiredCount(requiredCount_);
+    }
+
+    /**
+     * @dev Before calling this function to deposit protocol tokens, 
+     * users need approve this contract to be able to spend or transfer their protocol tokens.
+     * @param _amount amount of protocol tokens deposited to this contract
+     */
+    function deposit(uint256 _amount) public {
+        require(_amount > 0, "amount should larger than zero");
+
+        IERC20(protocolTokenAddress).safeTransferFrom(msg.sender, address(this), _amount);
+
+        emit Deposit(msg.sender, _amount);
+    }
+
+    /**
+     * @param _signs array includes signs from authorized signers, amount/length should equal to requiredCount
+     * @param _recipient address of recipient
+     * @param _contractAddress addresss of destination contract (should be this contract address)
+     * @param _amount amount of tokens to be withdrawed
+     * @param _nonce nonce to avoid replay attacks
+     * @param _signs array includes signs from authorized signers
+     */
+    function withdraw(address _recipient, address _contractAddress, uint256 _amount, uint256 _nonce, bytes[] memory _signs) public {
+        require(_contractAddress == address(this), "contract address disagreement");
+        require(_amount > 0, "amount should larger than zero");
+
+        uint256 balance = IERC20(protocolTokenAddress).balanceOf(address(this));
+        require(balance >= _amount, "balance not enough");
+
+        uint256 oldNonce = userNonce[msg.sender];
+        require(_nonce > oldNonce, "nonce smaller than last know nonce");
+        userNonce[msg.sender] = _nonce;
+
+        bytes32 _hash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", keccak256(abi.encodePacked(_recipient, _contractAddress, _amount, _nonce))));
+        
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        address signer;
+        bytes memory sign;
+        bytes[] memory signs = new bytes[](requiredCount);
+        address[] memory verifiedSigners = new address[](requiredCount);
+
+        signs = _signs;
+
+        for (uint256 i = 0; i < requiredCount; i++) {
+            sign = signs[i];
+            require(sign.length == 65, "length of sign is wrong");
+
+            assembly {
+                r := mload(add(sign, 0x20))
+                s := mload(add(sign, 0x40))
+                v := add(0x1b, byte(0, mload(add(sign, 0x60))))
+            }
+
+            signer = ecrecover(_hash, v, r, s);
+            require(isSigner[signer], "signer not authorized");
+
+            for (uint256 q = 0; q < i; q++) {
+                require(signer != verifiedSigners[q], "signer verified");
+            }
```

Comment:
you could sort signatures off-chain so that `verifiedSigners` will be a sorted array. there is little difference if the array is short, of course.